### ISSUE TITLE
How many connections mailcrab can open?

### ISSUE THREAD

**Issue by jokaorgua** [tweedegolf/mailcrab#49]
I have a strange situation when using mailcrab docker image with php and symfony. 

I’ve started a mailcrab container, symfony consumer container and symfony cli container. 

When consumer is started I can not open a connection from cli container. I get request timeout. If I stop the consumer container everything is working from cli. 

It seems that consumer container connects to mailcrab and makes its port busy so mailcrab can not open more connections for other clients at the same time. 

Mailcrab has only one thread? Or this is only mine problem ?

thanks

**Comment by stappersg** [comment]
On Tue, Apr 11, 2023 at 10:24:55AM -0700, jokaorgua wrote:

> Or this is only mine problem ?

@jokaorgua Please help answering that question
by documenting how to reproduce the "problem".

**Comment by jokaorgua** [comment]
I’m not sure if you are familiar with symfony consumers. 
1. Make a container with mailcrab
2. Make another container with symfony consumer which will consume async queue and send emails. Start it. 
3. Start one more container with bash
4. Try to open port 1025 with telnet from bash container and send couple of commands directly to mailcrab. 

I guess that section 2 can be replaced with cli container where connection to mailcrab will be left opened. But I’m not sure as I haven’t tested such a way.

**Comment by stappersg** [comment]
On Tue, Apr 11, 2023 at 10:39:11AM -0700, jokaorgua wrote:
> I’m not sure if you are familiar with symfony consumers.

I'm sure it is not fair to expect such tooling at hand.


> But I’m not sure as I haven’t tested such a way. 

FWIW:  My setup doesn't currently work, something I have to fix first.
(I have clue / idea  
     but have other things planned for the next 48hours)



Regards
Geert Stappers
Who wouldn't be surprised if "single thread is a Docker thing"

**Comment by marlonbaeten** [comment]
Currently Mailcrab uses [mailin](https://crates.io/crates/mailin-embedded) to handle SMTP connections. Mailin does use a threadpool to handle connections. By default it starts up [4 threads](https://gitlab.com/alienscience/mailin/-/blob/master/mailin-embedded/src/lib.rs#L73) This means Mailcrab can handle at most 4 connections concurrently, so make sure to properly close any SMTP connections when they are finished.
If you use docker compose could you post your `docker-compose.yml` - your problem might be in the network configuration.

**Comment by stappersg** [comment]
On Tue, Apr 11, 2023 at 11:23:46PM -0700, jokaorgua wrote:
> I've made a minimal docker-compose which can reproduce a problem


I can reproduce the reported issue without docker-compose:

  *  cargo run backend
  *  go to second terminal
  *  telnet 127.0.0.1 1025
  *  SMTP banner appears
  *  go to third terminal
  *  telnet 127.0.0.1 1025
  *  no SMTP banner
  *  go back to second terminal
  *  type `quit` and enter  ( to end SMTP session )
  *  go back to third terminal
  *  do see that the SMTP banner appeared


Groeten
Geert Stappers
Has no idea what could cause the blocking
-- 
Silence is hard to parse

**Comment by jokaorgua** [comment]
@stappersg thanks for narrowing the problem. so it exists not only for me

**Comment by stappersg** [comment]
On Tue, Apr 11, 2023 at 11:51:07PM -0700, jokaorgua wrote:
> @stappersg thanks for narrowing the problem. so it exists not only for me

Insult accepted.

**Comment by marlonbaeten** [comment]
Check - I can reproduce the issue.
I've narrowed the problem down to the threaded connection handling in mailin and created a fix: https://gitlab.com/alienscience/mailin/-/merge_requests/30
When I run Mailcrab with the patch applied to mailin it handles connections concurrently.
Once this is merged upstream I will release a new version of Mailcrab containing the fix.
